api/handlers.go

```diff
@@ -67,6 +69,55 @@ func (api *API) limitsGET(w http.ResponseWriter, _ *http.Request, _ httprouter.P
 
 // loginPOST starts a user session by issuing a cookie
 func (api *API) loginPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	err := req.ParseForm()
+	if err != nil {
+		api.staticLogger.Log(logrus.WarnLevel, "Error parsing POST form:", err)
+	}
+	email := req.PostFormValue("email")
+	pw := req.PostFormValue("password")
+	if email != "" && pw != "" {
+		api.loginPOSTCredentials(w, req, email, pw)
+		return
+	}
+	api.loginPOSTToken(w, req)
+}
+
+// loginPOSTCredentials is a helper that handles logins with credentials.
+func (api *API) loginPOSTCredentials(w http.ResponseWriter, req *http.Request, email, password string) {
+	// Fetch the user with that email, if they exist.
+	u, err := api.staticDB.UserByEmail(req.Context(), email, false)
+	if err != nil {
+		api.staticLogger.Tracef("Error fetching a user with email '%s': %+v\n", email, err)
+		api.WriteError(w, err, http.StatusUnauthorized)
+		return
+	}
+	// Check if the password matches.
+	err = hash.Compare(password, []byte(u.PasswordHash))
```

Comment:
nit: should we do this first to avoid a db query if the password is wrong?

Reply:
We are comparing the password to the hash we get from the DB. If we can do that without fetching the data from the DB first we're gonna revolutionize the entire computer world. :P

Reply:
oh ha sorry missed that 🙃